L1 Track Extended Collection for Phase2OT added to Validation Package

brandiskip opened the pull request:
#### PR description:

In Validation/SiTrackerPhase2V, added code to create necessary histograms using extended collection so that we now have histograms for both prompt and displaced tracks. Moved all code related to stubs to Validation/SiTrackerPhase2V/plugins/Phase2OTValidateStub.cc and created Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestStub.cc to keep all stub related code together and separate from L1Tracks code. Renamed necessary files to better describe what is being done in that particular code. Summary of histograms added for extended collection can be found 
[here](https://github.com/user-attachments/files/22659809/DQMOfflineSoftwareMeeting_Updated_Histograms.pdf).


Also included in this PR, code changes that ljuckett made to create histograms that include vectors for easier reading and remove redundancy.

In DQM/SiTrackerPhase2, fixed bug related to PS strip histogram, see slides 2 and 3 of this [talk](https://indico.cern.ch/event/1555122/contributions/6548463/attachments/3080726/5453258/DQMOfflineSoftwareMeeting_6_4_25.pdf).

#### PR validation:

scram build code-checks
scram build code-format

brandiskip committed:
added d0 cut

brandiskip committed:
added additional selection cuts

brandiskip committed:
added comments to top of file

brandiskip committed:
added InputTag and module for Extended Tracks

brandiskip committed:
restore Phase2OTHarvestTracks and Phase2OTValidateTracks, delete Reconstruction versions

brandiskip committed:
reduced number of files running on

brandiskip committed:
changed geom to D110 but after error, went back to D98

brandiskip committed:
changed number of stats

brandiskip committed:
removed trackingParticleExtendedValidOT to prevent calling analyze() twice

brandiskip committed:
add file as part of this CMSSW release

brandiskip committed:
fixed bug related to PS modules strip side hists

brandiskip committed:
added computeZ0LxyD0 helper function using TrackingParticle

brandiskip committed:
added computeZ0LxyD0 helper function using TrackingParticle

brandiskip committed:
fixed bug from running scram build code-checks

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"00c593dbebfdb423ab4ff6874d72d133b099e024":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestReconstruction.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateReconstruction.cc"],"squashed":false,"time":1740411545},"023fd99e0f2d1515859fb4a6d8a7344c4c15c3f8":{"files":["DQM/SiTrackerPhase2/test/harvestingstep_phase2tk_cfg.py"],"squashed":false,"time":1741714540},"099449278029d6c9595789f557a0ec051ff7d4c6":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTrackingParticles.cc","Validation/SiTrackerPhase2V/python/Phase2TrackerValidationFirstStep_cff.py","Validation/SiTrackerPhase2V/python/Phase2ValidateL1TTObjects_cff.py"],"squashed":false,"time":1737123001},"0b6adc369ea2ed1b6e2fd6733677ee94f4ddee4d":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateReconstruction.cc"],"squashed":false,"time":1747925817},"0dc02feeee0b9af738ed84fbf03719fb2d76c2a5":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestStub.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc","Validation/SiTrackerPhase2V/python/Phase2TrackerMCHarvesting_cff.py"],"squashed":false,"time":1747855446},"0f606612add792c0a58a541cc5b922c3e665e45c":{"files":["DQM/SiTrackerPhase2/test/dqmstep_phase2tk_cfg.py"],"squashed":false,"time":1741714500},"1256d65e36d8d3b26db637c83cbd40f3dc0e6e63":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestReconstruction.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateReconstruction.cc"],"squashed":false,"time":1747854005},"13f352ffd15fe0847a267ac770c996488c387532":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestReconstruction.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateReconstruction.cc"],"squashed":false,"time":1741728058},"1cc63092c5a26599650dea8070525fe2618dd267":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestReconstruction.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateReconstruction.cc","Validation/SiTrackerPhase2V/python/Phase2TrackerMCHarvesting_cff.py","Validation/SiTrackerPhase2V/python/Phase2ValidateL1TTObjects_cff.py"],"squashed":false,"time":1738344073},"1e8cdc0bf4d66b198a9b2c5889376c32c21f2bc0":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateReconstruction.cc"],"squashed":false,"time":1743971161},"1e9e0065298d135f9295e00a105f69f6d0570723":{"files":["DQM/SiTrackerPhase2/test/dqmstep_phase2tk_cfg.py"],"squashed":false,"time":1747853895},"219537f7c9adfdcad2b14f1cc261a16f9decaeda":{"files":["Validation/SiTrackerPhase2V/interface/TrackerPhase2ValidationUtil.h"],"squashed":false,"time":1749229902},"267aa4f2e16d6c7c38487c0af1744091e348c28c":{"files":["DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc","DQM/SiTrackerPhase2/python/Phase2TrackerMonitorDigi_cff.py","DQM/SiTrackerPhase2/python/Phase2TrackerMonitorDigi_cfi.py"],"squashed":false,"time":1749229451},"28b474bf5b5dd1201699f314d3b36e40751be79d":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestReconstruction.cc"],"squashed":false,"time":1738344294},"3153ac177fd26a10519fbd0b8b814738b222bbac":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTrackingParticles.cc"],"squashed":false,"time":1737133286},"346d3982682d7ae378c2cbd1ee2cd6d0504c5727":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestTrackingParticles.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTrackingParticles.cc"],"squashed":false,"time":1735588166},"3645d8427963a03075037467bfdaa403ccde9151":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestTrackingParticles.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTrackingParticles.cc"],"squashed":false,"time":1735849608},"382edbe0ef07f197612cfc9099c2a1667dc68a56":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc"],"squashed":false,"time":1745944009},"3b95f9cc0373912b858e19b1c4b0ff0950bd2396":{"files":["DQM/SiTrackerPhase2/test/dqmstep_phase2tk_cfg.py"],"squashed":false,"time":1741728081},"4100152f318f4d6b60d1d453d3685cb1a061f933":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestReconstruction.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateReconstruction.cc"],"squashed":false,"time":1741717712},"7a654393723f37c3a7989421c589fcb9a4faa920":{"files":["DQM/SiTrackerPhase2/test/dqmstep_phase2tk_cfg.py","DQM/SiTrackerPhase2/test/harvestingstep_phase2tk_cfg.py"],"squashed":false,"time":1736869126},"86eb242d9f8a6c1ee0c05be3ec0ebf52820e0f4e":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc"],"squashed":false,"time":1734705250},"8a0289f1fd7e9704f26ca0d486002cacb66c06cd":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc"],"squashed":false,"time":1741714668},"920961bbc2f30418ce014658e36301abb786b10a":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateReconstruction.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc"],"squashed":false,"time":1749229991},"a70b36242545b1c9fa5937ce04dc8b8555098143":{"files":["Validation/SiTrackerPhase2V/interface/TrackerPhase2ValidationUtil.h"],"squashed":false,"time":1749474568},"aaa69c062720fa5fbb6074ce4bc6805f6f01899c":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestReconstruction.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateReconstruction.cc"],"squashed":false,"time":1743528359},"afe31eb8bcb9f2f7aa93de401689055ddd5a1e89":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTrackingParticles.cc"],"squashed":false,"time":1734543197},"b592328a461e382f6a19de2194139068966f3d48":{"files":["DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc","Validation/SiTrackerPhase2V/interface/TrackerPhase2ValidationUtil.h","Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestTracks.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateStub.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTracks.cc"],"squashed":false,"time":1749471052},"bb1f81839dfcf7fda0f144e65c0232ff7af79c80":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateReconstruction.cc","Validation/SiTrackerPhase2V/python/Phase2TrackerValidationFirstStep_cff.py","Validation/SiTrackerPhase2V/python/Phase2ValidateL1TTObjects_cff.py"],"squashed":false,"time":1742831576},"bdf2331146c5a9ae773d89955f384113a3a87004":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc"],"squashed":false,"time":1736869160},"bfa568a21ac4a4ca66e45b27f0353f1331f2c6f8":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc"],"squashed":false,"time":1734540382},"c842cb46b5e98c1409842a93f58e94209a2fe8df":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc"],"squashed":false,"time":1736344160},"cc54c0a8355c1fa2d37e7029e68f11aabf51d145":{"files":["Validation/SiTrackerPhase2V/python/HLTPhase2TrackerValidationFirstStep_cff.py"],"squashed":false,"time":1743524038},"ccb2f172ad76ca2d45bc41142d09f8c484153924":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc"],"squashed":false,"time":1743524001},"ceb05463abf3eb54614ec4bc642cd89180848595":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTrackingParticles.cc"],"squashed":false,"time":1737990740},"d33f3989b52adbd0057acad380d0b1d588048d69":{"files":["Validation/SiTrackerPhase2V/BuildFile.xml"],"squashed":false,"time":1749229675},"e20f07eada196988290690261f7eefd7cc1d711d":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateReconstruction.cc"],"squashed":false,"time":1748017669},"e50047bc2df03fce496ccc288632e7b5aeb672e7":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestTracks.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateStub.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTracks.cc","Validation/SiTrackerPhase2V/python/Phase2TrackerMCHarvesting_cff.py","Validation/SiTrackerPhase2V/python/Phase2ValidateL1TTObjects_cff.py"],"squashed":false,"time":1749474676},"e733d784e52ee28384fc7cc13d35bea08d98145e":{"files":["DQM/SiTrackerPhase2/test/dqmstep_phase2tk_cfg.py","Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestReconstruction.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateReconstruction.cc"],"squashed":false,"time":1744648305},"e979a4f862e958349fed6f7da89eccfc14dd9417":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestTracks.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateStub.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTracks.cc"],"squashed":false,"time":1749469882},"f0783c2d1f38a5c6a7c44964fcf5a92223ad872a":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateReconstruction.cc"],"squashed":false,"time":1741877716},"f3ad69b8fa67b4c6de5589b6910915228cf0a4bd":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestReconstruction.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateReconstruction.cc"],"squashed":false,"time":1741714600},"f8b87b78706084145152f4d6f13e2548e82f4d13":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateReconstruction.cc"],"squashed":false,"time":1745944047}},"emoji":{},"last_seen_sha":"e50047bc2df03fce496ccc288632e7b5aeb672e7","signatures":{}} -->

mmusich commented on Validation/SiTrackerPhase2V/python/HLTPhase2TrackerValidationFirstStep_cff.py:
```diff
@@ -0,0 +1,37 @@
+from Validation.SiTrackerPhase2V.Phase2OTValidateCluster_cff import *
+from Validation.SiTrackerPhase2V.Phase2ITValidateRecHit_cff import *
+from Validation.SiTrackerPhase2V.Phase2ITValidateTrackingRecHit_cff import *
+from Validation.SiTrackerPhase2V.Phase2OTValidateTrackingRecHit_cff import *
```
why is this file being duplicated? 
See https://github.com/cms-sw/cmssw/blob/master/Validation/SiTrackerPhase2V/python/HLTPhase2TrackerValidationFirstStep_cff.py

tomalin commented:
Please change title to mention "L1 track" and "P2".
Can you also in the PR description provide a link where one can see the full set of plots being added?

brandiskip committed:
changed to CMSSW_15_1_0_pre3 release and made necessary code edits

brandiskip commented:
> Please change title to mention "L1 track" and "P2". Can you also in the PR description provide a link where one can see the full set of plots being added?

@tomalin These tasks have been completed.

brandiskip committed:
changed pT efficiency ymax from 2.0 to 1.0

tomalin commented on Validation/SiTrackerPhase2V/plugins/Phase2OTValidateStub.cc:
```diff
@@ -482,23 +517,171 @@ void Phase2OTValidateTTStub::analyze(const edm::Event& iEvent, const edm::EventS
+      continue;
+    if (tmp_tp_pt < TP_minPt)
+      continue;
+    if (tmp_tp_Lxy > 1.0)
```
Magic numbers not allowed

tomalin commented on Validation/SiTrackerPhase2V/interface/TrackerPhase2ValidationUtil.h:
```diff
@@ -1,12 +1,54 @@
+    float charge = tp.charge();
+
+    float t = std::tan(2.0 * std::atan(1.0) - 2.0 * std::atan(std::exp(-eta)));
+    float delx = -vx;
```
Given that delx and dely are only used once and their definition is very short, I'd suggest directly using -vx and -vy instead.

tomalin commented on Validation/SiTrackerPhase2V/interface/TrackerPhase2ValidationUtil.h:
```diff
@@ -1,12 +1,54 @@
+    if (delphi > pi)
+      delphi -= 2.0 * pi;
+
+    float z0 = vz + t * delphi / (2.0 * K);
```
Is this just a more accurate calculation of z0 than tp.z0() -- which has a comment in TrackingParticle.h saying that it is approximate. If you've validated it, we should perhaps just replace the calculation of TrackingParticle::z0(), so you could use that function, and everyone else could also profit from a more accurate result.

brandiskip replied:
This is the same calculation as in L1TrackNtupleMaker.cc. These calculations were previously part of [Phase2OTValidateReconstruction.cc](https://github.com/cms-sw/cmssw/blob/master/Validation/SiTrackerPhase2V/plugins/Phase2OTValidateReconstruction.cc#L296), I just moved this to a utility function as I use it in two different code.

tomalin commented on Validation/SiTrackerPhase2V/interface/TrackerPhase2ValidationUtil.h:
```diff
@@ -1,12 +1,54 @@
+    float x0p = delx - A * std::sin(phi);
+    float y0p = dely + A * std::cos(phi);
+    const float pi = 4.0 * std::atan(1.0);
+    float delphi = phi - std::atan2(-K * x0p, K * y0p);
```
Just #include DataFormats/Math/interface/deltaPhi.h and then do float deltaphi = reco::deltaPhi(phi, std::atan2(-K*x0p, K*y0p). Then it is automatically in the range -PI to +PI.

tomalin commented on DQM/SiTrackerPhase2/python/Phase2TrackerMonitorDigi_cff.py:
```diff
@@ -86,13 +86,22 @@
         ymax = 2.5,
         switch = True
+    ),
+    PositionOfDigisSHPS = digiMon.PositionOfDigisSHPS.clone(
```
I thought clone() allowed you to only specify the parameters that will be different to the original?

brandiskip replied:
Yes, fixed the issue where `clone` was being used unnecessarily

tomalin commented on Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestTracks.cc:
```diff
@@ -0,0 +1,1344 @@
+    resPt1->SetMinimum(0.0);
+    resPt1->SetStats(false);
+
+    for (int i = 0; i < 6; i++) {
```
Do I understand that we have a separate histogram Pt resolution for each bin a 2D (eta,Pt) space, and that the harvester determines the RMS width of each of these histograms, and uses it to plot Pt resolution as a function of eta and Pt? If it were necessary to do a Gaussian fit to get the resolution, I could understand this. But this seems a bit over the top. We should try to keep this code as simple and short as possible, so that it is easy to understand and maintain. Are TProfile histograms supported by the DQM? Would a profile histogram of fabs(reco - true) have done the trick, possibly scaling by sqrt(pi/2) to correct for the expect difference between RMS and abs(deviation)?

brandiskip replied:
Discussed this in a [BES meeting](https://indico.cern.ch/event/1566618/contributions/6599023/attachments/3109405/5511576/L1Tk_Meeting_Update_7_23_25.pdf) and it was decided to keep full histograms as is.

tomalin commented:
Is the number of histograms being added here and the L1 of detail motivated by looking at similar P1 histograms? e.g. Did they have lots of plots of resolution in helix parameters in 2D bins of eta and Pt? Do we expect bugs to only show up if we bin this finely? And is this number reasonable for the validator who has to check them (given that they must also check all the other Tracker histograms)?

brandiskip committed:
fixed magic number

brandiskip committed:
fixed magic numbers and cleanup

brandiskip committed:
reduced several lines of code by use of utility function

brandiskip committed:
added utility function to reduce lines of code in HarvestTracks and HarvestStub

brandiskip committed:
removed unnecessary use of clone()

brandiskip committed:
reduced several lines of code by use of utility function

brandiskip committed:
updated arguments from utility function computeZ0LxyD0

brandiskip committed:
fixed bug for cluster position hists for PS modules

brandiskip committed:
resolve conflict: missing lines of code

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48275/45559

ERROR: Unable to merge PR.

See log https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48275/45559/cms-checkout-topic.log

brandiskip committed:
Merge branch 'master' into testing_changes

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48275/45560

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48275/45560/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48275/45560/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

brandiskip committed:
changed formatting for code-checks

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48275/45561

cmsbuild commented:
Pull request #48275 was updated.

brandiskip committed:
update to CMSSW_15_1_0_pre6 and run code checks

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48275/46038

ERROR: Unable to merge PR.

See log https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48275/46038/cms-checkout-topic.log

brandiskip committed:
Merge branch 'master' into testing_changes

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48275/46039

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48275/46039/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48275/46039/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_16_0_X. Please open a backport if it should also go in to CMSSW_15_1_X.

brandiskip commented on Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestStub.cc:
```diff
@@ -0,0 +1,189 @@
+                                                          numerator->GetXaxis()->GetXmax());
+
+    // Calculate the efficiency
+    me_effic_clus_barrel->getTH1F()->Divide(numerator, denominator, 1., 1., "B");
```
Created a [header file](https://github.com/brandiskip/cmssw/blob/testing_changes/Validation/SiTrackerPhase2V/interface/TrackerPhase2HistUtil.h) to reduce these lines of code

mmusich replied:
Just FYI, such a class already exists [DQMGenericClient](https://github.com/cms-sw/cmssw/blob/master/DQMServices/Components/plugins/DQMGenericClient.cc).

brandiskip committed:
re-run code checks

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48275/46092

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48275/46092/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48275/46092/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

brandiskip committed:
Apply cms-bot code-format patch

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48275/46093

cmsbuild commented:
Pull request #48275 was updated.

brandiskip commented on Validation/SiTrackerPhase2V/interface/TrackerPhase2ValidationUtil.h:
```diff
@@ -1,12 +1,54 @@
   bool isPrimary(const SimTrack& simTrk, const PSimHit* simHit);
   static constexpr float cmtomicron = 1e4;
 
+  inline std::tuple<float, float, float> computeZ0LxyD0(const TrackingParticle& tp, float pt, float eta, float phi) {
```
Fixed

cmsbuild commented:
+1

**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-47e817/48294/summary.html
**COMMIT**: eea57ba0952333ae4ab61705ea086003976ecb4a
**CMSSW**: CMSSW_16_0_X_2025-09-26-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/48275/48294/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 3 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_0_X_2025-09-26-1100+47e817/70513/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_0_X_2025-09-26-1100+47e817/70513/validateJR.html)
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3911159
 - DQMHistoTests: [Total failures: 111](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_0_X_2025-09-26-1100+47e817/70513/validateJR.html)
 - DQMHistoTests: Total nulls: 181
 - DQMHistoTests: Total successes: 3910847
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 265073.05 KiB( 50 files compared)
 - DQMHistoSizes: changed ( 29634.0,... ): 52518.150 KiB    TrackerPhase2ITDigi/DigiMonitor
 - DQMHistoSizes: changed ( 29634.0,... ): 443.875 KiB      TrackerPhase2OTDigi/DigiMonitor
 - DQMHistoSizes: changed ( 29634.0,... ): 51.081 KiB       TrackerPhase2OTL1TrackV/Extended_L1TF
 - DQMHistoSizes: changed ( 29634.0,... ): 25.482 KiB       TrackerPhase2OTL1TrackV/Nominal_L1TF
 - DQMHistoSizes: changed ( 29634.0,... ): -21.150 KiB      TrackerPhase2OTL1TrackV/ResolutionIngredients
 - DQMHistoSizes: changed ( 29634.0,... ): -5.277 KiB       TrackerPhase2OTL1TrackV/EfficiencyIngredients
 - DQMHistoSizes: changed ( 29634.0,... ): 2.164 KiB        TrackerPhase2OTStubV/EfficiencyIngredients
 - DQMHistoSizes: changed ( 29634.0,... ): 0.285 KiB        TrackerPhase2OTStubV/Residual
 - Checked 218 log files, 188 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_0_X_2025-09-26-1100+47e817/70513/triggerResults): no differences found

tomalin commented:
> Is the number of histograms being added here and the L1 of detail motivated by looking at similar P1 histograms? e.g. Did they have lots of plots of resolution in helix parameters in 2D bins of eta and Pt? Do we expect bugs to only show up if we bin this finely? And is this number reasonable for the validator who has to check them (given that they must also check all the other Tracker histograms)? Are the histograms similar in number and names to those used for the prompt L1 tracks? If not why not? If they are, then could code for booking and filling many of the histograms be shared between prompt and displaced tracking?

Hi @brandiskip is there any answer to this comment? Also, is the picture of the histograms added by this PR still given by https://indico.cern.ch/event/1555122/contributions/6548463/attachments/3080726/5461274/DQM_Extended_Collection_Histograms.pdf ?

brandiskip commented:
> Is the number of histograms being added here and the L1 of detail motivated by looking at similar P1 histograms? e.g. Did they have lots of plots of resolution in helix parameters in 2D bins of eta and Pt? Do we expect bugs to only show up if we bin this finely? And is this number reasonable for the validator who has to check them (given that they must also check all the other Tracker histograms)? Are the histograms similar in number and names to those used for the prompt L1 tracks? If not why not? If they are, then could code for booking and filling many of the histograms be shared between prompt and displaced tracking?

@tomalin Sorry, I thought I replied to this. I haven't looked at P1 histograms. As far as reasonable amount of histograms, this is something we need to understand better. All the new histograms I added are the same as prompt L1 tracks that already existed when I started except for d0, which I added. For filling hists, I do have shared code for each as seen [here](https://github.com/brandiskip/cmssw/blob/testing_changes/Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTracks.cc#L440-L471). For booking, the code could probably be reduced but I'll have to think about this. 

>Also, is the picture of the histograms added by this PR still given by [https://indico.cern.ch/event/1555122/contributions/6548463/attachments/3080726/5461274/DQM_Extended_Collection_Histograms.pdf](https://indico.cern.ch/event/1555122/contributions/6548463/attachments/3080726/5461274/DQM_Extended_Collection_Histograms.pdf) ?

Yes, these are the histograms being added by this PR.

gabrielmscampos commented on Validation/SiTrackerPhase2V/plugins/Phase2OTValidateStub.cc:
```diff
@@ -482,23 +513,170 @@ void Phase2OTValidateTTStub::analyze(const edm::Event& iEvent, const edm::EventS
       }
       // Fill the appropriate histogram based on layer/disc
-      if (layer >= 1 && layer <= trklet::N_LAYER) {
+      if (layer >= 1 && layer <= 6) {
```
Perhaps I'm missing something, but why was this changed from `trklet::N_LAYER` to `6`, similar to other parts of the code that `trklet::N_DISK` was changed to `5`?

brandiskip replied:
This was an error. I'm not sure what happened but it has been resolved.

brandiskip committed:
Replaced magic numbers with N_LAYER and N_DISK

brandiskip committed:
fix code format issues

brandiskip commented on Validation/SiTrackerPhase2V/interface/TrackerPhase2HistUtil.h:
```diff
@@ -0,0 +1,33 @@
+#include "TH1.h"
+
+namespace phase2tkutil {
+  inline void makeEfficiencyME(TH1* numerator,
+                               TH1* denominator,
+                               dqm::legacy::MonitorElement* me,
+                               const std::string& xAxisTitle) {
+    TH1* efficiency = me->getTH1();
+    efficiency->Divide(numerator, denominator, 1., 1., "B");
+    efficiency->SetMinimum(0.0);
+    efficiency->SetMaximum(1.1);
+    efficiency->SetStats(false);
+    efficiency->GetXaxis()->SetTitle(xAxisTitle.c_str());
+    efficiency->GetYaxis()->SetTitle("Efficiency");
+  }
```
I want this PR to get merged as is but this is something I will look into for a future PR.

gabrielmscampos commented:
+dqm
